nomad/job_endpoint_hooks.go

```diff
@@ -307,45 +308,95 @@ func (v *jobValidate) Validate(job *structs.Job) (warnings []error, err error) {
-			if serviceErr != nil {
-				multierror.Append(validationErrors, serviceErr)
-			}
-			if len(serviceWarn) > 0 {
-				warnings = append(warnings, serviceWarn...)
-			}
+			serviceWarns, serviceErrs := v.validateServiceIdentity(s, fmt.Sprintf("task group %s", tg.Name))
+			multierror.Append(validationErrors, serviceErrs)
+			warnings = append(warnings, serviceWarns...)
 		}
 
 		for _, t := range tg.Tasks {
 			for _, s := range t.Services {
-				serviceWarn, serviceErr := v.validateServiceIdentity(s)
-				if serviceErr != nil {
-					multierror.Append(validationErrors, serviceErr)
-				}
-				if len(serviceWarn) > 0 {
-					warnings = append(warnings, serviceWarn...)
-				}
+				serviceWarns, serviceErrs := v.validateServiceIdentity(s, fmt.Sprintf("task %s", t.Name))
+				multierror.Append(validationErrors, serviceErrs)
+				warnings = append(warnings, serviceWarns...)
 			}
+
+			vaultWarns, vaultErrs := v.validateVaultIdentity(t)
+			multierror.Append(validationErrors, vaultErrs)
+			warnings = append(warnings, vaultWarns...)
 		}
 	}
 
 	return warnings, validationErrors.ErrorOrNil()
 }
 
-func (v *jobValidate) validateServiceIdentity(s *structs.Service) (warnings []error, err error) {
+func (v *jobValidate) validateServiceIdentity(s *structs.Service, parent string) (warnings []error, err error) {
+	var mErr *multierror.Error
+
+	// Prefix errors and warnings.
+	defer func() {
+		prefix := fmt.Sprintf("Service %s in %s", s.Name, parent)
+		for i, w := range warnings {
+			warnings[i] = fmt.Errorf("%s %s", prefix, w)
+		}
+		err = multierror.Prefix(mErr, prefix)
+	}()
+
 	if s.Identity != nil {
 		if !v.srv.config.UseConsulIdentity() {
-			return nil, fmt.Errorf("service %s defines an identity but server configuration for consul.use_identity is not true", s.Name)
+			mErr = multierror.Append(mErr, errors.New("defines an identity but server is not configured to use Consul identities, set use_identity to true in the Consul server configuration"))
 		}
 
 		if s.Identity.Name == "" {
-			return nil, fmt.Errorf("identity for service %s has an empty name", s.Name)
+			mErr = multierror.Append(mErr, errors.New("has an identity with an empty name"))
 		}
 	} else if v.srv.config.UseConsulIdentity() && v.srv.config.ConsulServiceIdentity() == nil {
-		return nil, fmt.Errorf("service %s does not have an identity and no default service identity is provided", s.Name)
+		mErr = multierror.Append(mErr, errors.New("expected to have an identity, add an identity block to the service or provide a default using the service_identity block in the server Consul configuration"))
+	}
+
+	return
+}
+
+func (v *jobValidate) validateVaultIdentity(t *structs.Task) (warnings []error, err error) {
+	var mErr *multierror.Error
+
+	// Prefix errors and warnings.
+	defer func() {
+		prefix := fmt.Sprintf("Task %s", t.Name)
+		for i, w := range warnings {
+			warnings[i] = fmt.Errorf("%s %s", prefix, w)
+		}
+		err = multierror.Prefix(mErr, prefix)
+	}()
```

Comment:
I promise I tried to make these errors not 💩, but the whole `multierror` and `[]error` thing makes it very hard to generate a good output. I started to refactor this whole error system to give it more structure, but it snowballed pretty quickly.

Here's a warning output example:

![image](https://github.com/hashicorp/nomad/assets/775380/fb6b2669-9dea-4308-96a9-07037022b1ff)